src/codegen/encoding/borsh_encoding.rs

```diff
@@ -693,4 +717,584 @@ impl BorshEncoding {
+    }
+
+    /// Decodes a complex array from a borsh encoded buffer
+    fn decode_complex_array(
+        &self,
+        array_var: &Expression,
+        buffer: &Expression,
+        offset_var: usize,
+        offset_expr: &Expression,
+        dimension: usize,
+        elem_ty: &Type,
+        dims: &[ArrayLength],
+        validator: &mut BufferValidator,
+        ns: &Namespace,
+        vartab: &mut Vartable,
+        cfg: &mut ControlFlowGraph,
+        indexes: &mut Vec<usize>,
+    ) {
+        // If we have a 'int[3][4][] vec', we can only validate the buffer after we have
+        // allocated the outer dimension, i.e., we are about to read a 'int[3][4]' item.
+        if validator.validation_necessary()
+            && !dims[0..(dimension + 1)]
+                .iter()
+                .any(|d| *d == ArrayLength::Dynamic)
+        {
+            let mut elems = BigInt::one();
+            for item in &dims[0..(dimension + 1)] {
+                elems.mul_assign(item.array_length().unwrap());
+            }
+            elems.mul_assign(elem_ty.memory_size_of(ns));
+            let elems_size = Expression::NumberLiteral(Loc::Codegen, Type::Uint(32), elems);
+            validator.validate_offset_plus_size(offset_expr, &elems_size, vartab, cfg);
+            validator.validate_array();
+        }
+
+        // Dynamic dimensions mean that the subarray we are processing must be allocated in memory.
+        if dims[dimension] == ArrayLength::Dynamic {
+            let offset_to_validate = increment_four(offset_expr.clone());
+            validator.validate_offset(offset_to_validate, vartab, cfg);
+            let array_length = retrieve_array_length(buffer, offset_expr, vartab, cfg);
+            cfg.add(
+                vartab,
+                Instr::Set {
+                    loc: Loc::Codegen,
+                    res: offset_var,
+                    expr: increment_four(offset_expr.clone()),
+                },
+            );
+            let new_ty = Type::Array(Box::new(elem_ty.clone()), dims[0..(dimension + 1)].to_vec());
+            let allocated_array = allocate_array(&new_ty, array_length, vartab, cfg);
+
+            if indexes.is_empty() {
+                if let Expression::Variable(_, _, var_no) = array_var {
+                    cfg.add(
+                        vartab,
+                        Instr::Set {
+                            loc: Loc::Codegen,
+                            res: *var_no,
+                            expr: Expression::Variable(
+                                Loc::Codegen,
+                                new_ty.clone(),
+                                allocated_array,
+                            ),
+                        },
+                    );
+                } else {
+                    unreachable!("array_var must be a variable");
+                }
+            } else {
+                // TODO: This is wired up for multidimensional dynamic arrays, but they do no work yet
```

Comment:
Is there a warning or anything for the user during compilation, or do you plan to fix this soon?

Reply:
No. Multidimensional arrays are broken in Solang. We have an issue to make it work: #932.

Reply:
I'll add a link to the issue on the comment to make it clear.